### PR TITLE
Use the real lit tool for lit testing

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
For historical reasons, we've used a custom shell script to run our lit
tests. We've made it less bad over time, but it's time to just use the
real thing. This gets the real lit tool running under Bazel (and CMake,
but that's less hard).

I hope to upstream some of this Bazel stuff, but LLVM's lit setup is
way more complicated, so checking this in here for now.

Fixes https://github.com/google/iree/issues/818

### CODE REVIEW

File: iree/samples/vision/CMakeLists.txt [hunk 1]
```diff
@@ -43,7 +43,7 @@ iree_lit_test(
     iree_run_mnist_module_test
   TEST_FILE
     "mnist_test.txt"
-  DATA
+  TOOLS
```
Comment (ScottTodd) [inline iree/samples/vision/CMakeLists.txt:46]: new failure in here?

```
FAIL: IREE :: samples/vision/mnist_test.txt (1 of 1)
******************** TEST 'IREE :: samples/vision/mnist_test.txt' FAILED ********************
Script:
--
: 'RUN: at line 1';   (iree-run-mnist-module) | FileCheck /tmpfs/src/github/iree/iree/samples/vision/mnist_test.txt
--
Exit Code: 2

Command Output (stderr):
--
../iree/tools/utils/image_util.c:44: NOT_FOUND; can't load image mnist_test.png; load image
FileCheck error: '<stdin>' is empty.
FileCheck command line:  FileCheck /tmpfs/src/github/iree/iree/samples/vision/mnist_test.txt
```

logs: https://source.cloud.google.com/results/invocations/29a1c095-42e7-4504-bb96-42cd58c88626/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log

File: build_tools/cmake/iree_lit_test.cmake [hunk 2]
```diff
@@ -52,10 +50,17 @@ function(iree_lit_test)
...
+  foreach(_DATA_DEP IN LISTS _RULE_DATA _RULE_TOOLS)
     list(APPEND _DATA_DEP_PATHS $<TARGET_FILE:${_DATA_DEP}>)
-  endforeach(_DATA_DEP)
+  endforeach()
+
+  set(_LIT_PATH_ARGS)
+  # TODO: only tools
+  foreach(_TOOL IN LISTS _RULE_TOOLS)
+    list(APPEND _LIT_PATH_ARGS "--path" "$<TARGET_FILE_DIR:${_TOOL}>")
+  endforeach()
```
Comment (ScottTodd) [inline build_tools/cmake/iree_lit_test.cmake:63]: TODO still relevant? Add a name and a bit more context?

File: iree/samples/vision/CMakeLists.txt (hunk 1 above)
Comment (GMNGeoffrey) [reply iree/samples/vision/CMakeLists.txt:46]: Yeah I think I have samples turned off in my local build, so I missed this. And it isn't covered by Bazel->CMake

File: build_tools/cmake/iree_lit_test.cmake (hunk 2 above)
Comment (GMNGeoffrey) [reply build_tools/cmake/iree_lit_test.cmake:63]: Thanks for catching. I've gotten in the habit of sticking bare TODOs in to remind me to fix something before submission/sending for review, but we don't have the presubmit here that yells at you for bare TODOs :-D

**Comment by GMNGeoffrey** [comment]
Ok got things fixed and cleaned up some of our FileCheck dependence a bit as well. Now with real lit we get things like placeholders for the current directory as well as the current file :partying_face:

File: iree/tools/CMakeLists.txt [hunk 3]
```diff
@@ -459,12 +459,11 @@ if(${IREE_BUILD_COMPILER})
...
-    # Ensure FileCheck gets built. Tests don't have dependencies in CMake
-    # because they aren't targets. So until we fix that, we just force this to
-    # get built.
-    set_target_properties(FileCheck PROPERTIES EXCLUDE_FROM_ALL OFF)
-  endif()
+  # Ensure FileCheck gets built. Tests don't have dependencies in CMake because
+  # they aren't targets. So until we fix that, we just force this to get built.
+  # Limiting this to when IREE_BUILD_TESTS is set prevents the installation
+  # below, which we use for cross-platform testing.
+  set_target_properties(FileCheck PROPERTIES EXCLUDE_FROM_ALL OFF)
```
Comment (ScottTodd) [inline iree/tools/CMakeLists.txt:466]: Can you trigger https://buildkite.com/iree/iree-build-configurations on this PR? I want to verify that https://github.com/google/iree/blob/main/build_tools/cmake/build_runtime.sh doesn't build LLVM code

**Review by ScottTodd — APPROVED** [review]

File: iree/tools/CMakeLists.txt (hunk 3 above)
Comment (GMNGeoffrey) [reply iree/tools/CMakeLists.txt:466]: Yes, automerge is doing that for you by merging this :-P

Comment (ScottTodd) [reply iree/tools/CMakeLists.txt:466]: Ah, automerge.. well, that will run on postsubmit anyways :P I can keep an eye on https://buildkite.com/iree/iree-build-configurations/builds/1887


aaand as I'm typing this it already finished. Looks good.

**Pull request merged**